internal/rpcproxy/keymanager.go

```diff
@@ -0,0 +1,132 @@
+	keys, err := s.ListUserPublicKeys(ctx, s.user)
+	if err != nil {
+		return jujuparams.StringsResults{}, err
+	}
+
+	var formatter func(key sshkeys.PublicKey) string
+	switch args.Mode {
+	case ssh.FullKeys:
+		formatter = marshalAuthorizedKeyWithComment
+	case ssh.Fingerprints:
+		formatter = fingerprintWithComment
+	default:
+		return jujuparams.StringsResults{}, fmt.Errorf("unknown mode (%v)", args.Mode)
+	}
+
+	// The Juju CLI takes the first element from the slice of stringsResults.
+	res := jujuparams.StringsResult{}
+	for _, key := range keys {
+		res.Result = append(res.Result, formatter(key))
+	}
+	return jujuparams.StringsResults{Results: []jujuparams.StringsResult{res}}, nil
+}
+
+// AddKeys saves the SSH keys defined in args and associates them
+// with the authenticated user.
+func (s *keyManagerFacade) AddKeys(ctx context.Context, args jujuparams.ModifyUserSSHKeys) (jujuparams.ErrorResults, error) {
+	var res []jujuparams.ErrorResult
+	errF := func(err error, msg string) jujuparams.ErrorResult {
+		return jujuparams.ErrorResult{Error: &jujuparams.Error{
+			Code:    string(errors.ErrorCode(err)),
+			Message: fmt.Sprintf("%s: %s", msg, err.Error()),
+		}}
+	}
+
+	for i, key := range args.Keys {
+		out, comment, _, _, err := gossh.ParseAuthorizedKey([]byte(key))
+		if err != nil {
+			res = append(res, errF(err, fmt.Sprintf("Failed to parse key (entry %d)", i)))
+		}
+		jimmKey := sshkeys.PublicKey{
+			PublicKey: out,
+			Comment:   comment,
+		}
+		if err := s.AddUserPublicKey(ctx, s.user, jimmKey); err != nil {
+			res = append(res, errF(err, fmt.Sprintf("Failed to add key (comment %s)", comment)))
+		}
+	}
+
+	return jujuparams.ErrorResults{Results: res}, nil
+}
+
+// DeleteKeys removes saved keys associated with the authenticated user
+// and finds keys to remove either by comment or fingerprint.
+func (s *keyManagerFacade) DeleteKeys(ctx context.Context, args jujuparams.ModifyUserSSHKeys) (jujuparams.ErrorResults, error) {
+	var res []jujuparams.ErrorResult
+	errF := func(err error, msg string) jujuparams.ErrorResult {
+		return jujuparams.ErrorResult{Error: &jujuparams.Error{
+			Code:    string(errors.ErrorCode(err)),
+			Message: fmt.Sprintf("%s: %s", msg, err.Error()),
+		}}
+	}
+
+	for _, key := range args.Keys {
+		if isFingerprintRegexp.MatchString(key) {
+			err := s.RemoveUserKeyByFingerprint(ctx, s.user, key)
+			if err != nil {
+				res = append(res, errF(err, fmt.Sprintf("Failed to remove key by fingerprint (%s)", key)))
+			}
+		} else {
+			err := s.RemoveUserKeyByComment(ctx, s.user, key)
+			if err != nil {
+				res = append(res, errF(err, fmt.Sprintf("Failed to remove key by comment (%s)", key)))
+			}
+		}
+	}
+
+	return jujuparams.ErrorResults{Results: res}, nil
+}
+
+func marshalAuthorizedKeyWithComment(key sshkeys.PublicKey) string {
+	// Copied from gossh.MarshalAuthorizedKey with an addition for the comment.
+	// Errors from the buffer's Write..() methods are always nil.
```

Comment:
Move this up to the actual godoc please, also I've forogtten - what is the comment for?

Reply:
An SSH key doesn't have a comment, but the OpenSSH authorized key format for public keys does and Juju stores the key comment and displays it back to the user when they list keys, helping to identify for a user where the key came from.